### PR TITLE
Adding props to Checkbox component

### PR DESCRIPTION
**Pull request by xman343**
## Changes

Adding the option for customizable wrapper props to be added to the Checkbox component. Issue: https://github.com/iTwin/iTwinUI/issues/1368
## Testing

React unit testing and visual testing undertaken.
## Docs

Docs will be updated to outline wrapperProps utility.

### CODE REVIEW

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx [hunk 1]
```diff
@@ -92,18 +109,24 @@ export const Checkbox = React.forwardRef((props, ref) => {
             'iui-checkbox-visibility': variant === 'eyeball',
             'iui-loading': isLoading,
           },
-          className && { [className]: !label },
+          boxClassName && { [boxClassName]: !label },
```
Comment (gretanausedaite) [inline packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:112]: I think you got confused a bit: this is the main Checkbox component - the little square, therefore, it should get "main" className, style, etc. props. 
Undo this change and check if you are able to customise `label`.

Comment (xman343) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:112]: Isn't label already being customized by WrapperProps?   wrapperProps?: React.ComponentProps<'label'>;
Or do you mean the label as in the text?

**Comment by xman343** [comment]
Just remembered I need to update the entry in Docs - I'll do that very soon!

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx [hunk 2]
```diff
@@ -104,6 +109,12 @@ export const Checkbox = React.forwardRef((props, ref) => {
     </>
   );
 
+  const {
+    className: wrapperClassName,
+    style: wrapperStyle,
```
Comment (adamhock) [inline packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:114]: Since we're not doing any manipulation to `wrapperStyle` before we add it to the wrapper element, this doesn't need to be extracted. `style` customization will be included when `...restWrapperProps` is added to the wrapper element

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx [hunk 3]
```diff
@@ -116,9 +127,10 @@ export const Checkbox = React.forwardRef((props, ref) => {
           [`iui-${status}`]: !!status,
           'iui-loading': isLoading,
         },
-        className,
+        wrapperClassName,
```
Comment (adamhock) [inline packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:130]: Now that the class name for the label comes from `wrapperClassName`, there is a check in the `checkbox` element that only adds `className` to the checkbox when there is no label. This check can be removed and `className` can be added directly to the checkbox element regardless of if there is a label or not

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx [hunk 4]
```diff
@@ -116,9 +127,10 @@ export const Checkbox = React.forwardRef((props, ref) => {
...
       )}
-      style={style}
+      style={wrapperStyle}
+      {...restWrapperProps}
     >
       {checkbox}
       {label && (
```
Comment (adamhock) [inline packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:136]: @gretanausedaite Should we do the same thing here for label as we did for the wrapper, or is passing a custom `React.ReactNode` for `label` enough customization?

File: packages/itwinui-react/src/core/Checkbox/Checkbox.test.tsx [hunk 5]
```diff
@@ -187,6 +189,19 @@ it('renders correctly with visibility checkbox', () => {
   expect(container.querySelector('.iui-checkbox-visibility')).toBeTruthy();
 });
 
+// wrapperProps test
```
Comment (adamhock) [inline packages/itwinui-react/src/core/Checkbox/Checkbox.test.tsx:192]: This comment can be removed

**Review by adamhock — COMMENTED** [review]
Make sure you add a changeset and link the issue number that this PR is referencing in the PR description

**Comment by xman343** [comment]
> Make sure you add a changeset and link the issue number that this PR is referencing in the PR description

Done

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx [hunk 6]
```diff
@@ -86,14 +90,10 @@ export const Checkbox = React.forwardRef((props, ref) => {
...
-            'iui-checkbox-visibility': variant === 'eyeball',
-            'iui-loading': isLoading,
-          },
-          className && { [className]: !label },
-        )}
+        className={cx('iui-checkbox', {
+          'iui-checkbox-visibility': variant === 'eyeball',
+          'iui-loading': isLoading,
+        })}
         style={!label ? style : undefined}
```
Comment (gretanausedaite) [inline packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:97]:
```suggestion
        style={style}
```

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx (hunk 4 above)
Comment (gretanausedaite) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:126]: I think it's enough. label className simply sets line-height, everything else is up to the user.
@mayank99 agree?

Comment (mayank99) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:126]: no. every single element needs to be customizable.

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx (hunk 3 above)
Comment (adamhock) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:121]: `className` is still needed for users to add a custom class name to `iui-checkbox`. Only the check for `label` needed to be removed

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx [hunk 7]
```diff
@@ -36,6 +36,14 @@ type CheckboxProps = {
...
    */
   isLoading?: boolean;
+  /**
+   * Passes properties for checkbox wrapper.
+   */
+  wrapperProps?: React.ComponentProps<'label'>;
+  /**
+   * Passes properties for checkbox label.
+   */
+  labelProps?: React.ComponentProps<'label'>;
```
Comment (adamhock) [inline packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:46]: Since these props are added to a `span` element, make sure this prop's type changes from `<'label'>` to `<'span'>`

**Review by adamhock — COMMENTED** [review]
Once we get visual and unit tests to pass, I will approve

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx (hunk 7 above)
Comment (xman343) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx]: Done!

**Comment by xman343** [comment]
Image changes were added. Visual React test in CI won't pass without them, but I don't think the image changes originated from my PR. Any advice? 
87e89bc0-1d6e-494c-a7e4-61b87a3aa479

File: .changeset/empty-rockets-try.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+---
+'@itwin/itwinui-react': major
+---
+
+Added wrapperProps to allow for styling of CheckBox wrapper.
```
Comment (gretanausedaite) [inline .changeset/empty-rockets-try.md:5]: Update changeset with all changes you've made

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx [hunk 9]
```diff
@@ -116,13 +130,20 @@ export const Checkbox = React.forwardRef((props, ref) => {
...
+      {...restWrapperProps}
     >
       {checkbox}
       {label && (
-        <Box as='span' className='iui-checkbox-label'>
+        <Box
+          as='span'
+          className={
+            (cx('iui-checkbox-label'),
+            labelClassName ? labelClassName : className)
```
Comment (gretanausedaite) [inline packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:143]: With this code, if user has labeled checkbox and adds `custom-classname`, but not labelClassName/wrapperClassName, html structure would be this:
image

File: .changeset/empty-rockets-try.md (hunk 8 above)
Comment (xman343) [reply .changeset/empty-rockets-try.md]: Done

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx (hunk 9 above)
Comment (xman343) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:143]: I wrote it like that so that the `label` and `wrapper` would share a className with the parent component by default, unless the user goes out of their way to specify custom classNames for `label` and `wrapper`. Should their classNames both be different by default?

Comment (gretanausedaite) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:143]: Why they need to share className?

Comment (xman343) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:143]: They don't have to, I just figured some users might want them to. If they shouldn't share classNames by default, then I'll change the code so they don't

Comment (gretanausedaite) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:143]: If user wants to add styles to certain part of component, they need to use props for that part of component.
Remove this check.

Comment (xman343) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx]: Removed the checks.

Two unit tests are now failing again. I think it's because they're assuming that setting the `className` variable for the component would still assign it to both the `input` and the `label`, when this isn't the case anymore. I could rewrite the tests so that they make sure the label *isn't* automatically assigned the same `className` as the `input` when only the `input` is given the assignment. 
image
image

Comment (mayank99) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx]: i think those two tests can be adjusted/removed because the behavior that they were testing has been changed significantly.

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx [hunk 10]
```diff
@@ -116,13 +130,17 @@ export const Checkbox = React.forwardRef((props, ref) => {
...
       )}
-      style={style}
+      {...restWrapperProps}
     >
       {checkbox}
       {label && (
-        <Box as='span' className='iui-checkbox-label'>
+        <Box
+          as='span'
+          className={(cx('iui-checkbox-label'), labelClassName)}
```
Comment (mayank99) [inline packages/itwinui-react/src/core/Checkbox/Checkbox.tsx:141]: there is a huge (yet subtle) typo in this code

```suggestion
          className={cx('iui-checkbox-label', labelClassName)}
```

this should fix your issue above with tests and you should be able to revert images
> Image changes were added. Visual React test in CI won't pass without them, but I don't think the image changes originated from my PR. Any advice?

this also makes me realize you are missing a unit test for this important functionality.

Comment (xman343) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx]: Good eye! Fixed the typo. 

> this also makes me realize you are missing a unit test for this important functionality.

What sort of unit test should I ad?

Comment (mayank99) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx]: same as the one for wrapperProps. test the behavior you are adding.

Comment (xman343) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx]: Added.

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx (hunk 9 above)
Comment (xman343) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx]: Modified` "Should isomorphically apply class on %s"` to only test `input`, and also removed the final failing test.

Comment (mayank99) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx]: it is no longer isomorphic so it's not testing anything special and can be removed

imo you should have a single unit test that tests adding separate custom classes all three elements (`className`, `labelProps`, `wrapperProps`)

Comment (xman343) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx]: Done

Comment (mayank99) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx]: you're only testing two cases in there, missing a test for `className` which gets applied on the input.

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx (hunk 10 above)
Comment (mayank99) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx]: now update images too

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx (hunk 9 above)
Comment (xman343) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx]: Added

File: packages/itwinui-react/src/core/Checkbox/Checkbox.tsx (hunk 10 above)
Comment (xman343) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.tsx]: Done!

File: packages/itwinui-react/src/core/Checkbox/Checkbox.test.tsx [hunk 11]
```diff
@@ -115,6 +114,7 @@ it.each(['label', 'input'] as const)(
     const { container } = render(
       <Checkbox
         label={el === 'label' ? 'Some label' : undefined}
+        wrapperProps={{ style: { color: 'blue' } }}
```
Comment (adamhock) [inline packages/itwinui-react/src/core/Checkbox/Checkbox.test.tsx:117]: I think instead of having this `it.each` statement, this test can be written similarly to how your test for `className` is written. Just apply different stylings to `wrapperProps`, `labelProps`, and the input, and check for those three stylings within this test.

It might even be possible to test for `style` and `className` within the same test labeled `'correctly passes props through wrapperProps and labelProps`.

Also, I would make sure you apply different colors to each style instead of `blue` across the board.

**Review by adamhock — COMMENTED** [review]
A few nitpicky comments about the unit tests, but other than that, things look good to me!

File: packages/itwinui-react/src/core/Checkbox/Checkbox.test.tsx (hunk 11 above)
Comment (xman343) [reply packages/itwinui-react/src/core/Checkbox/Checkbox.test.tsx]: Done!

File: .changeset/empty-rockets-try.md [hunk 12]
```diff
@@ -0,0 +1,5 @@
+---
+'@itwin/itwinui-react': major
+---
+
+Added wrapperProps and labelProps to allow for styling of CheckBox wrapper and label elements.
```
Comment (mayank99) [inline .changeset/empty-rockets-try.md:5]: Worth mentioning the breaking change in className/style behavior.

```suggestion
Checkbox: `className` and `style` will now always be applied on the checkbox input element. Added new `wrapperProps` and `labelProps` to allow for styling of wrapper and label elements.
```

**Review by mayank99 — APPROVED** [review]

**Review by adamhock — APPROVED** [review]
Once the changeset description changes, it's ready to go! 🎉 Great work

**Comment by xman343** [comment]
> Once the changeset description changes, it's ready to go! 🎉 Great work

done!
